godot-core/src/builtin/plane.rs

```diff
@@ -191,6 +310,301 @@ mod test {
+        assert!(plane_a.is_equal_approx(&plane_a));
+        assert_eq!(plane_a.intersect_3(&plane_a, &plane_a), None);
+        // Two coincident planes with an unrelated third plane would intersect in every point along the
+        // intersection line, thus no unique solution.
+        assert!(plane_b.is_equal_approx(&plane_b));
+        assert_eq!(plane_b.intersect_3(&plane_b, &plane_g), None);
+        // Two coincident planes with a parallel third plane would have no common intersection.
+        assert_eq!(plane_a.intersect_3(&plane_a, &plane_d), None);
+        // Three planes that intersects each other in a common line would intersect in every point along
+        // the line, thus no unique solution.
+        assert_eq!(plane_a.intersect_3(&plane_f, &plane_g), None);
+        // Three planes that intersects each other in 3 parallel lines would not intersect in a common
+        // point.
+        assert_eq!(plane_h.intersect_3(&plane_i, &plane_j), None);
+    }
+
+    /// Tests `intersects_ray()`.
+    #[test]
+    fn test_ray_intersections() {
+        // Plane that is flat along the z-axis.
+        let plane: Plane = Plane::new(Vector3::new(0.0, 0.0, 1.0).normalized(), 0.0);
+
+        // Origin point.
+        let vec_a: Vector3 = Vector3::new(0.0, 0.0, 0.0);
+        // Forms a straight line along the z-axis with `vec_a` that is perpendicular to `plane`.
+        let vec_b: Vector3 = Vector3::new(0.0, 0.0, 0.5);
+        let vec_c: Vector3 = Vector3::new(0.0, 0.0, 1.0);
+        let vec_d: Vector3 = Vector3::new(0.0, 0.0, -1.0);
+        // Forms a slanted line with `vec_a` relative to `plane`.
+        let vec_e: Vector3 = Vector3::new(0.5, 0.5, 0.0);
+        // Forms a line with `vec_c` that is parallel with `plane`.
+        let vec_f: Vector3 = Vector3::new(1.0, 0.0, 1.0);
+
+        // From a point straight up from the origin point, a ray pointing straight down would cross
+        // the `plane` in the origin point.
+        assert_eq!(plane.intersects_ray(vec_b, vec_d), Some(vec_a));
+        assert_eq!(plane.intersects_ray(vec_c, vec_d), Some(vec_a));
+        // From a point straight down the origin point, a ray pointing straight up would cross the `plane`
+        // in the origin point.
+        assert_eq!(plane.intersects_ray(vec_d, vec_b), Some(vec_a));
+        assert_eq!(plane.intersects_ray(vec_d, vec_c), Some(vec_a));
+        // From a point straight up the origin point, the distance from the origin to the intersection point
+        // formed by a slanted ray should follow the Pythagorean Theorem.
+        assert_eq!(
+            plane.intersects_ray(vec_c, vec_e - vec_c).unwrap().length(),
+            (vec_c.distance_squared_to(vec_e) - vec_c.length_squared()).sqrt()
+        );
+        assert_eq!(
+            plane.intersects_ray(vec_b, vec_e - vec_b).unwrap().length(),
+            (vec_b.distance_squared_to(vec_e) - vec_b.length_squared()).sqrt()
+        );
```

Comment:
Well, I thought it would be useful to check for mathematically accurate results for the implementations too since it's a mathematical object 😅, but I can see why not comparing straight from the results of `intersect_ray` (already removed the s) would be problematic for just testing for implementation. 

But I don't understand one thing, so should I remove the `.length()` (the derived property I assume) or just scrap the whole Pythagorean Theorem thing (or maybe the relation to the Pythagorean Theorem is the derived property)?

I can tell I'm probably reading too much into this.